src/app/dao/tests/JoinerDAO.spec.ts

```diff
@@ -1,17 +1,462 @@
+
+            // Then it should fail
+            await expectPermissionToBeDenied(result);
+        });
+        it('should allow to change other fields than candidates and partStatus to STARTED', async() => {
+            // Given a user that created a (non-started) part and joiner
+            const partId: string = (await createPart({ createJoiner: true })).partId;
+
+            const updates: Partial<Joiner>[] = [
+                { chosenOpponent: UserMocks.CANDIDATE_MINIMAL_USER },
+                { partStatus: PartStatus.CONFIG_PROPOSED.value },
+                { firstPlayer: FirstPlayer.CHOSEN_PLAYER.value },
+                { partType: PartType.BLITZ.value },
+                { maximalMoveDuration: 73 },
+                { totalPartDuration: 1001001 },
+            ];
+            for (const update of updates) {
+                // When modifying a field
+                const result: Promise<void> = joinerDAO.update(partId, update);
+
+                // Then it should succeed
+                await expectAsync(result).toBeResolvedTo();
+            }
+        });
+        describe('on started part', () => {
+            let partId: string;
+            beforeEach(async() => {
+                // Given a part that is started
+
+                // Creator creates the joiner
+                const createdPartId: string = (await createPart({ createJoiner: true, signOut: true })).partId;
+                partId = createdPartId;
+
+                // A candidate adds themself to the candidates list
+                const candidate: MinimalUser = await addCandidate(partId);
+
+                // The creator then selects candidates as the chosen opponent and proposes the config
+                await reconnectUser(CREATOR_EMAIL);
+                const update: Partial<Joiner> = {
+                    chosenOpponent: candidate,
+                    partStatus: PartStatus.CONFIG_PROPOSED.value,
+                };
+                await expectAsync(joinerDAO.update(partId, update)).toBeResolvedTo();
+                await signOut();
+
+                // The chosen opponent then accepts the part
+                await reconnectUser(CANDIDATE_EMAIL);
+                const result: Promise<void> = joinerDAO.update(partId, { partStatus: PartStatus.PART_STARTED.value });
+                await expectAsync(result).toBeResolvedTo();
+
+                // And we act as the creator
+                await reconnectUser(CREATOR_EMAIL);
+            });
+            it('should allow to change partStatus to FINISHED after it is STARTED', async() => {
+                // When changing partStatus to FINISHED
+                const result: Promise<void> = joinerDAO.update(partId, { partStatus: PartStatus.PART_FINISHED.value });
+
+                // Then it should succeed
+                await expectAsync(result).toBeResolvedTo();
+            });
+            it('should forbid to change fields after part has started', async() => {
```

Comment:
neither creator, that's already said that it's a set-only-once, but still, let's be explicit!

Reply:
ajouté !